### PR TITLE
Remove the SyntaxData abstraction layer

### PR DESCRIPTION
**Pull request by ahoppen**
`SyntaxData` just held the data of the `Syntax` node and didn’t provide any additional functionality. Instead, it required unecessary casts between `Syntax` and `SyntaxData` and we were inconsistent about where we took `Syntax` or `SyntaxData`. Store the members of `SyntaxData` directly in `Syntax`.

I saw slight (0.5% - 2%) performance improvements in instruction count when running swift-format with this change.

The best way to review this is commit-by-commit. The first two commits just split up files, the last commit is the actual removal of `SyntaxData`.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

**Review by natikgadzhi — COMMENTED** [review]
Ah, thank you! I've learned a lot reading through the changes, and I think this change simplifies things.

A few nitpicks and traces of `SyntaxData`, but it looks like the code should work fine to me.

File: CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxBaseNodesFile.swift [hunk 1]
```diff
@@ -73,7 +73,7 @@ let syntaxBaseNodesFile = SourceFileSyntax(leadingTrivia: copyrightHeader) {
           // We know this cast is going to succeed. Go through init(_: SyntaxData)
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxBaseNodesFile.swift:73]: Good catch. Thanks 👍🏽

**Comment by ahoppen** [comment]
@natikgadzhi You are right about the documentation. I’d prefer to keep these changes out of this PR though because it’s already quite big.

**Review by CodaFi — APPROVED** [review]

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

File: Sources/SwiftSyntax/SyntaxIdentifier.swift [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+/// nodes in the tree, not just the modified node's children.
+public struct SyntaxIdentifier: Hashable {
+  /// Unique value for the root node.
+  ///
+  /// Multiple trees may have the same 'rootId' if their root RawSyntax is the
+  /// same instance. This guarantees that the trees with the same 'rootId' have
+  /// exact the same structure. But, two trees with exactly the same structure
+  /// might still have different 'rootId's.
+  let rootId: UInt
+  /// Unique value for a node within its own tree.
```
Comment (kimdv) [inline Sources/SwiftSyntax/SyntaxIdentifier.swift:68]: Some places do we have a newline after declarations. Personally I think it's easier to read. 
What do you think of adding new lines after a declaration ?

**Review by kimdv — COMMENTED** [review]
Added some inline comments up for discussion

File: Sources/SwiftSyntax/SyntaxProtocol.swift [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+    return withExtendedLifetime(self) {
+      return Syntax(raw: self.raw, rawNodeArena: self.raw.arena).cast(Self.self)
+    }
+  }
+}
+
+// MARK: Type information
+
+extension SyntaxProtocol {
+  /// The kind of the syntax node, e.g. if it is a `functionDecl`.
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SwiftSyntax/SyntaxProtocol.swift:92]: Good idea. I just copied that comment but might as well improve it.

File: Sources/SwiftSyntax/SyntaxProtocol.swift [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+  /// at the dynamic kind of this node and returns ``FunctionDeclSyntax``.
+  public var syntaxNodeType: SyntaxProtocol.Type {
+    return self.raw.kind.syntaxNodeType
+  }
+}
+
+// MARK: Children / parent
+
+public extension SyntaxProtocol {
+  /// A sequence over the `present` children of this node.
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SwiftSyntax/SyntaxProtocol.swift:114]: Oh, that comment still pre-dated the introduction of `SyntaxTreeViewMode`. I removed the `present` part of the comment.

File: Sources/SwiftSyntax/SyntaxIdentifier.swift [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  public static func < (lhs: SyntaxIndexInTree, rhs: SyntaxIndexInTree) -> Bool {
+    return lhs.indexInTree < rhs.indexInTree
+  }
+}
+
+/// Provides a stable and unique identity for ``Syntax`` nodes.
+///
+/// Note that two nodes might have the same contents even if their IDs are
+/// different. For example two different ``FunctionDeclSyntax`` nodes in the
+/// might have the exact same contents but if they occur at a different
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SwiftSyntax/SyntaxIdentifier.swift:53]: Removed “in the”

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test macOS

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Review by kimdv — APPROVED** [review]

**Review by bnbarham — APPROVED** [review]

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Pull request merged**